apm-agent-core/src/main/java/co/elastic/apm/agent/logging/LoggingConfiguration.java

```diff
@@ -298,34 +297,7 @@ public void assertValid(Boolean value) {
         .buildWithDefault(LogFormat.PLAIN_TEXT);
 
     public static void init(List<ConfigurationSource> sources, String ephemeralId) {
-        // The initialization of log4j may produce errors if the traced application uses log4j settings (for
-        // example - through file in the classpath or System properties) that configures specific properties for
-        // loading classes by name. Since we shade our usage of log4j, such non-shaded classes may not (and should not)
-        // be found on the classpath.
-        // All handled Exceptions should not prevent us from using log4j further, as the system falls back to a default
-        // which we expect anyway. We take a calculated risk of ignoring such errors only through initialization time,
-        // assuming that errors that will make the logging system non-usable won't be handled.
-        String initialListenersLevel = System.setProperty(INITIAL_LISTENERS_LEVEL, "OFF");
-        String initialStatusLoggerLevel = System.setProperty(INITIAL_STATUS_LOGGER_LEVEL, "OFF");
-        String defaultListenerLevel = System.setProperty(DEFAULT_LISTENER_LEVEL, "OFF");
-        try {
-            Configurator.initialize(new Log4j2ConfigurationFactory(sources, ephemeralId).getConfiguration());
-        } catch (Throwable throwable) {
-            System.err.println("[elastic-apm-agent] ERROR Failure during initialization of agent's log4j system: " + throwable.getMessage());
-        } finally {
-            restoreSystemProperty(INITIAL_LISTENERS_LEVEL, initialListenersLevel);
-            restoreSystemProperty(INITIAL_STATUS_LOGGER_LEVEL, initialStatusLoggerLevel);
-            restoreSystemProperty(DEFAULT_LISTENER_LEVEL, defaultListenerLevel);
-            StatusLogger.getLogger().setLevel(Level.ERROR);
-        }
-    }
-
-    private static void restoreSystemProperty(String key, @Nullable String originalValue) {
-        if (originalValue != null) {
-            System.setProperty(key, originalValue);
-        } else {
-            System.clearProperty(key);
-        }
+        ConfigurationFactory.setConfigurationFactory(new Log4j2ConfigurationFactory(sources, ephemeralId));
```

Comment:
What's the difference between `ConfigurationFactory.setConfigurationFactory` and `Configurator.initialize`? What triggered the error and what exactly did it resolve?

Reply:
`ConfigurationFactory` is a singleton that allows overriding its instance through `ConfigurationFactory.setConfigurationFactory` (not considered thread safe, but since we do it so early on when there is only the main thread it should be OK). Once you did that, any logger created thereafter will be configured through it.
`Configurator.initialize` only initializes a `LoggerContext`, so that all loggers within this context will implicitly get the same configuration. However, once you try to create a logger in a different context, it will attempt to get a configuration and will look it up from what it can find (e.g. properties/xml file).
The default context locator is based on class loader and now (> 1.26.0), when we have many class loaders that load classes that request loggers, only loggers within the same context as the initialized one (which is the agent class loader) will be properly configured and all the other contexts (eg plugin class loaders) won't.
There are multiple ways to handle that, like making sure we initialize all contexts or use a different context selector (will probably still require us to make sure we initialize multiple contexts, for example - threads), but I think the proper way is to utilize the fact that we have isolated logging system and so we can just make sure everything within this system is configured the same.

Reply:
Thanks, much clearer now!